### PR TITLE
[FLINK-7958][metrics] Reporters provide default delimiter

### PR DESCRIPTION
**Pull request by zentol**
Currently, the delimiter used for assembling the metric identifier is either based on a delimiter configured for the reporter, or the globally configured delimiter (default: '.').
Additionally (just for context), there are some instances around logicalScopes where reporters themselves can decide on the delimiter.

External systems usually have well-defined requirements on what the delimiter should be, and it thus seems reasonable to allow reporters to provide a default delimiter, instead of using a global default across all reporters.

This PR adds a method for retrieving this delimiter from the reporter. This delimiter is used as the default, and can be overwritten by explicitly configuring a delimiter for this reporter.

This means that the global delimiter is no longer used in the context of a reporter, which is a slight change in behavior that must be documented in the release notes.

While useful on it's own this PR is primarily a preparatory step for properly exposing the logical scope; which currently is tricky to do since reporters using this currently inject a non-default delimiter that isn't controllable. Due to not being controllable the used methods should not be exposed publicly, but this in turn requires a different mechanism for reporters to define a delimiter.

### CODE REVIEW

File: docs/_includes/generated/metric_configuration.html [hunk 1]
```diff
@@ -52,6 +52,11 @@
             <td style="word-wrap: break-word;">(none)</td>
             <td>The reporter interval to use for the reporter named &lt;name&gt;.</td>
         </tr>
+        <tr>
+            <td><h5>metrics.reporter.&lt;name&gt;.scope.delimiter</h5></td>
+            <td style="word-wrap: break-word;">(none)</td>
+            <td>The delimiter used to assemble the metric identifier the reporter named &lt;name&gt;.</td>
```
Comment (azagrebin) [inline docs/_includes/generated/metric_configuration.html]:
```suggestion
            <td>The delimiter used to assemble the metric identifier for the reporter named &lt;name&gt;.</td>
```

File: flink-core/src/main/java/org/apache/flink/configuration/MetricOptions.java [hunk 2]
```diff
@@ -52,16 +52,24 @@
...
 	public static final ConfigOption<String> REPORTER_INTERVAL =
 		key("metrics.reporter.<name>.interval")
 			.noDefaultValue()
 			.withDescription("The reporter interval to use for the reporter named <name>.");
 
+	// this option only exists for documentation purposes
+	public static final ConfigOption<String> REPORTER_SCOPE_DELIMITER =
+		key("metrics.reporter.<name>." + ConfigConstants.METRICS_REPORTER_SCOPE_DELIMITER)
+			.noDefaultValue()
+			.withDescription("The delimiter used to assemble the metric identifier the reporter named <name>.");
```
Comment (azagrebin) [inline flink-core/src/main/java/org/apache/flink/configuration/MetricOptions.java]:
```suggestion
			.withDescription("The delimiter used to assemble the metric identifier for the reporter named <name>.");
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricRegistryImpl.java [hunk 3]
```diff
@@ -147,7 +147,7 @@ public MetricRegistryImpl(MetricRegistryConfiguration config, Collection<Reporte
 					}
 					reporters.add(reporterInstance);
 
-					String delimiterForReporter = reporterSetup.getDelimiter().orElse(String.valueOf(globalDelimiter));
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricRegistryImpl.java]: It seems we continue to use the `globalDelimiter` in `AbstractMetricGroup` for some purposes.
Is it supposed to be so or is it another story with logical scopes?

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricRegistryImpl.java]: the abstract metric group can be used a) by reporters, in which case we rely on the reporter defined here, and b) user-functions, in which case they still rely on the global delimiter.

**Comment by zentol** [comment]
The delimiter is already exposed through the configuration, this feature isn't new.

As outlined in the PR description, this feature is primarily a preparation for follow-ups, namely exposing the logical scope via the MetricScope for which reporters need a way to define a non-default delimiter ('_') so that we can retain the current behavior of the prometheus/influx reporters. Similarly to your suggestion about the `CharacterFilter`, the delimiter is also mostly static, hence why I don't want to expose methods accepting a custom delimiter in the `MetricScope`.
Ironically, without this change the only way users would be able to have the correct delimiter is to configure it explicitly for the reporter.

**Comment by 1u0** [comment]
My point is that, in my opinion, this configuration (and also [the configuration for the default delimeter](https://github.com/apache/flink/pull/2219#discussion_r70475544)) is not in demand by users.
Instead of exposing this parameter outside of Flink to operators of Flink cluster, you can still make it configurable but internally. So when metrics register themselves, they would define which delimeter to use. End users probably won't need to customize this.

While in case of Prometheus reporter, `.` is [invalid character in metric name](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels). InfluxDB reporter just followed this convention (I don't remember exact reason why I didn't use `.`). It may worth to use `_` for all reporters (for unification, as some common ground), but I understand that this would be an incompatible change.

But thinking more generally, I find that the `(CharacterFilter, delimeter)` pair is just a concrete implementation of more generic api (something like `MetricScopeFormatter` that can become more specific (and different) to some metric reporters.

To summarise, my personal cons against this change are:
 * the configuration is not requested by users. Less configuration exposed by Flink - less things to maintain as public api and more freedom to change implementation later;
 * even if the config is exposed, actual correct values that would work in practice are limited (cannot be any character) and specific to actual metrics system;
 * for some reporters, the scope formatting can required to be more flexible (compared to current implementation based on `CharacterFilter` and `delimeter`).

**Comment by zentol** [comment]
Again, the config option _already exists_, this PR just _documents_ it. An _existing_ feature, that is _actually evaluated_ should be documented, **period**.

Removing the option is an _entirely_ separate discussion, and I do not understood why are focusing on this so much here.
